File: packages/esm-format-go/pkg/esm/closed_functions_test.go
package esm

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
)

// TestClosedFunctionsConformance drives every fixture under
// tests/closed_functions/<module>/<name>/{canonical.esm, expected.json}.
//
// For each fixture:
//  1. Load canonical.esm (exercises the parser's `fn`-op handling).
//  2. For each scenario in expected.json, evaluate the named closed
//     function with the given inputs and assert agreement with the
//     reference output within the declared tolerance.
//  3. For interp.searchsorted, drive `error_scenarios` and assert the
//     spec-pinned diagnostic code surfaces.
//
// All five bindings MUST run the same fixtures (esm-spec §9.4 / esm-aia).
func TestClosedFunctionsConformance(t *testing.T) {
	root := closedFunctionsFixturesDir(t)
	entries, err := walkClosedFnDirs(root)
	if err != nil {
		t.Fatalf("walk fixtures: %v", err)
	}
	if len(entries) == 0 {
		t.Fatalf("no closed-function fixtures under %s", root)
	}
	for _, e := range entries {
		e := e
		t.Run(e.label, func(t *testing.T) {
			// Step 1: load the canonical.esm. We only need the parser to
			// accept it — every fixture's RHS is a single `fn` invocation
			// of the function we're testing.
			if _, err := Load(e.canonical); err != nil {
				t.Fatalf("Load(%s): %v", e.canonical, err)
			}

			// Step 2 + 3: drive expected.json against EvaluateClosedFunction.
			raw, err := os.ReadFile(e.expected)
			if err != nil {
				t.Fatalf("read %s: %v", e.expected, err)
			}
			var fixture closedFnFixture
			if err := json.Unmarshal(raw, &fixture); err != nil {
				t.Fatalf("parse %s: %v", e.expected, err)
			}
			if fixture.Function == "" {
				t.Fatalf("%s: missing `function` field", e.expected)
			}
			if !IsClosedFunction(fixture.Function) {
				// Spec-first phased rollout (esm-94w and similar): a new
				// closed-function fixture lands in the spec PR before this
				// binding's implementation. Skip rather than fail; the
				// per-language [Impl] bead adds the function to the registry
				// and the fixture starts running automatically.
				t.Skipf("fixture function %q not yet implemented in this binding", fixture.Function)
			}
			for _, sc := range fixture.Scenarios {
				sc := sc
				t.Run("ok/"+sc.Name, func(t *testing.T) {
					args, err := normalizeFixtureArgs(sc.Inputs)
					if err != nil {
						t.Fatalf("normalize inputs: %v", err)
					}
					out, err := EvaluateClosedFunction(fixture.Function, args)
					if err != nil {
						t.Fatalf("EvaluateClosedFunction(%s, %v): %v", fixture.Function, args, err)
					}
					if err := assertCloseEnough(out, sc.Expected, fixture.Tolerance); err != nil {
						t.Fatalf("scenario %s (%s): %v", sc.Name, sc.Description, err)
					}
				})
			}
			for _, sc := range fixture.ErrorScenarios {
				sc := sc
				t.Run("err/"+sc.Name, func(t *testing.T) {
					args, err := normalizeFixtureArgs(sc.Inputs)
					if err != nil {
						t.Fatalf("normalize inputs: %v", err)
					}
					_, err = EvaluateClosedFunction(fixture.Function, args)
					if err == nil {
						t.Fatalf("expected error %q, got nil", sc.ExpectedErrorCode)
					}
					cfErr, ok := err.(*ClosedFunctionError)
					if !ok {
						t.Fatalf("expected *ClosedFunctionError, got %T: %v", err, err)
					}
					if cfErr.Code != sc.ExpectedErrorCode {
						t.Fatalf("error code mismatch: got %q, want %q (%s)",
							cfErr.Code, sc.ExpectedErrorCode, sc.Description)
					}
				})
			}
		})
	}
}

type closedFnFixture struct {
	Function       string                  `json:"function"`
	Tolerance      closedFnTolerance       `json:"tolerance"`
	Scenarios      []closedFnScenario      `json:"scenarios"`
	ErrorScenarios []closedFnErrorScenario `json:"error_scenarios"`
}

type closedFnTolerance struct {
	Abs float64 `json:"abs"`
	Rel float64 `json:"rel"`
}

type closedFnScenario struct {
	Name        string          `json:"name"`
	Description string          `json:"description"`
	Inputs      []interface{}   `json:"inputs"`
	Expected    json.RawMessage `json:"expected"`
}

type closedFnErrorScenario struct {
	Name              string        `json:"name"`
	Description       string        `json:"description"`
	Inputs            []interface{} `json:"inputs"`
	ExpectedErrorCode string        `json:"expected_error_code"`
}

type closedFnEntry struct {
	label     string // e.g. "datetime/year"
	canonical string
	expected  string
}

// walkClosedFnDirs returns every <module>/<function> directory under root
// that has both canonical.esm and expected.json.
func walkClosedFnDirs(root string) ([]closedFnEntry, error) {
	var out []closedFnEntry
	modules, err := os.ReadDir(root)
	if err != nil {
		return nil, err
	}
	for _, mod := range modules {
		if !mod.IsDir() {
			continue
		}
		modDir := filepath.Join(root, mod.Name())
		fns, err := os.ReadDir(modDir)
		if err != nil {
			return nil, err
		}
		for _, fn := range fns {
			if !fn.IsDir() {
				continue
			}
			fnDir := filepath.Join(modDir, fn.Name())
			canonical := filepath.Join(fnDir, "canonical.esm")
			expected := filepath.Join(fnDir, "expected.json")
			if _, err := os.Stat(canonical); err != nil {
				continue
			}
			if _, err := os.Stat(expected); err != nil {
				continue
			}
			out = append(out, closedFnEntry{
				label:     mod.Name() + "/" + fn.Name(),
				canonical: canonical,
				expected:  expected,
			})
		}
	}
	return out, nil
}

// normalizeFixtureArgs converts the JSON-decoded input slice into the
// types EvaluateClosedFunction expects: numbers stay as float64; nested
// arrays stay as []interface{}; the literal string "NaN" is converted to
// math.NaN(); other strings are forwarded for diagnostic purposes.
func normalizeFixtureArgs(inputs []interface{}) ([]interface{}, error) {
	out := make([]interface{}, len(inputs))
	for i, raw := range inputs {
		out[i] = normalizeFixtureValue(raw)
	}
	return out, nil
}

func normalizeFixtureValue(v interface{}) interface{} {
	switch x := v.(type) {
	case string:
		if x == "NaN" || x == "nan" || x == "NAN" {
			return math.NaN()
		}
		return x
	case []interface{}:
		out := make([]interface{}, len(x))
		for i, e := range x {
			out[i] = normalizeFixtureValue(e)
		}
		return out
	default:
		return v
	}
}

// assertCloseEnough compares a closed-function output to its expected
// reference within (abs OR rel) tolerance. Integer-typed outputs are
// promoted to float64 for the comparison; the spec pins zero tolerance
// for every entry except `datetime.julian_day` (≤ 1 ulp).
func assertCloseEnough(got interface{}, expectedRaw json.RawMessage, tol closedFnTolerance) error {
	want, err := decodeExpectedScalar(expectedRaw)
	if err != nil {
		return err
	}
	gf, ok := toFloat64(got)
	if !ok {
		return errf("got non-numeric scalar %T (%v)", got, got)
	}
	if math.IsNaN(want) {
		if math.IsNaN(gf) {
			return nil
		}
		return errf("expected NaN, got %g", gf)
	}
	diff := math.Abs(gf - want)
	if diff <= tol.Abs {
		return nil
	}
	if want != 0 && diff/math.Abs(want) <= tol.Rel {
		return nil
	}
	return errf("got %g, want %g (|diff|=%g, tol=abs %g rel %g)",
		gf, want, diff, tol.Abs, tol.Rel)
}

func decodeExpectedScalar(raw json.RawMessage) (float64, error) {
	s := strings.TrimSpace(string(raw))
	if s == "\"NaN\"" || s == "\"nan\"" || s == "\"NAN\"" {
		return math.NaN(), nil
	}
	var f float64
	if err := json.Unmarshal(raw, &f); err != nil {
		return 0, err
	}
	return f, nil
}

// errf is a tiny formatted-error helper used by the conformance harness.
func errf(format string, args ...interface{}) error {
	return fmt.Errorf(format, args...)
}

func closedFunctionsFixturesDir(t *testing.T) string {
	t.Helper()
	_, thisFile, _, _ := runtime.Caller(0)
	pkgDir := filepath.Dir(thisFile)
	// pkg/esm/closed_functions_test.go -> repo root is 4 levels up.
	repoRoot := filepath.Join(pkgDir, "..", "..", "..", "..")
	return filepath.Join(repoRoot, "tests", "closed_functions")
}

File: packages/EarthSciSerialization.jl/test/closed_functions_test.jl
# Closed function registry — Julia conformance harness adapter (esm-tzp / esm-4aw).
#
# Drives the cross-binding fixtures under `tests/closed_functions/<module>/<name>/`
# from the Julia binding: parse `canonical.esm` (validates the parser's `fn`-op
# handling), then walk the scenarios in `expected.json` and assert that
# `evaluate_closed_function` agrees with the reference output within the
# declared tolerance. The same fixture set runs from each binding's harness;
# any binding that disagrees with the spec-pinned values fails CI (esm-spec §9.4).

using Test
using JSON3
using EarthSciSerialization

const _CF_REPO_ROOT  = normpath(joinpath(@__DIR__, "..", "..", ".."))
const _CF_FIX_ROOT   = joinpath(_CF_REPO_ROOT, "tests", "closed_functions")

# Convert a JSON3-decoded scenario input to the value the closed function
# expects. Strings used as numeric placeholders (e.g. "NaN") are decoded;
# arrays recurse element-wise so `xs` arrays land as `Vector{Float64}`.
function _decode_input(v)
    if v isa String
        v == "NaN"  && return NaN
        v == "Inf"  && return Inf
        v == "-Inf" && return -Inf
        throw(ArgumentError("unrecognized string input: $(v)"))
    elseif v isa AbstractVector || v isa JSON3.Array
        return [_decode_input(x) for x in v]
    elseif v isa Bool
        throw(ArgumentError("boolean inputs not allowed"))
    elseif v isa Real
        return Float64(v)
    end
    throw(ArgumentError("unsupported input type: $(typeof(v))"))
end

# Tolerance comparison per esm-spec §9.2: pass if either |actual − expected|
# ≤ abs OR |actual − expected| ≤ rel·max(1, |expected|). The "max(1, ...)"
# guard avoids zero-relative tolerance when expected is zero.
function _within_tol(actual, expected, abs_tol, rel_tol)
    abs_tol = Float64(abs_tol); rel_tol = Float64(rel_tol)
    a = Float64(actual); e = Float64(expected)
    if isnan(a) && isnan(e)
        return true
    end
    diff = abs(a - e)
    return diff <= abs_tol || diff <= rel_tol * max(1.0, abs(e))
end

@testset "Closed function registry conformance (esm-tzp / esm-4aw)" begin
    @test isdir(_CF_FIX_ROOT)

    # Walk every <module>/<name> directory and run the scenarios it pins.
    for module_dir in sort(readdir(_CF_FIX_ROOT))
        full_module = joinpath(_CF_FIX_ROOT, module_dir)
        isdir(full_module) || continue
        @testset "$(module_dir)/*" begin
            for fname_dir in sort(readdir(full_module))
                fixture_dir = joinpath(full_module, fname_dir)
                isdir(fixture_dir) || continue
                canonical = joinpath(fixture_dir, "canonical.esm")
                expected  = joinpath(fixture_dir, "expected.json")
                @testset "$(fname_dir)" begin
                    @test isfile(canonical)
                    @test isfile(expected)

                    # Parser must accept the fixture (i.e. the `fn` op AST is
                    # valid under the v0.3.0 schema).
                    file = EarthSciSerialization.load(canonical)
                    @test file.esm == "0.3.0"

                    spec = JSON3.read(read(expected, String))
                    fn_name = String(spec.function)
                    if !(fn_name in closed_function_names())
                        # Spec-first phased rollout (esm-94w and similar): a
                        # new closed-function fixture lands in the spec PR
                        # before this binding's implementation. Skip rather
                        # than fail; the per-language [Impl] bead adds the
                        # function to the registry, at which point the
                        # fixture starts running automatically.
                        @info "skipping fixture $(fixture_dir): function $(fn_name) not yet implemented in this binding"
                        continue
                    end
                    abs_tol = haskey(spec, :tolerance) ? Float64(spec.tolerance.abs) : 0.0
                    rel_tol = haskey(spec, :tolerance) ? Float64(spec.tolerance.rel) : 0.0

                    for scenario in spec.scenarios
                        sname = String(scenario.name)
                        inputs_decoded = [_decode_input(v) for v in scenario.inputs]
                        actual = evaluate_closed_function(fn_name, inputs_decoded)
                        # Expected may also be a NaN/Inf string sentinel
                        # (esm-94w fixtures use this for nan-x / nan-y cases).
                        expected_val = _decode_input(scenario.expected)
                        @testset "$(sname)" begin
                            @test _within_tol(actual, expected_val, abs_tol, rel_tol)
                        end
                    end

                    # `error_scenarios` (when present) pin load-time / call-
                    # time error cases; the binding MUST raise a
                    # `ClosedFunctionError` whose `.code` field equals
                    # `expected_error_code`.
                    if haskey(spec, :error_scenarios)
                        for err in spec.error_scenarios
                            ename = String(err.name)
                            inputs_decoded = [_decode_input(v) for v in err.inputs]
                            expected_code = String(err.expected_error_code)
                            @testset "error: $(ename)" begin
                                err_caught = try
                                    evaluate_closed_function(fn_name, inputs_decoded)
                                    nothing
                                catch e
                                    e
                                end
                                @test err_caught isa ClosedFunctionError
                                if err_caught isa ClosedFunctionError
                                    @test err_caught.code == expected_code
                                end
                            end
                        end
                    end
                end
            end
        end
    end

    # Sanity: closed_function_names() returns the v0.3.0 set verbatim.
    @testset "closed_function_names() matches the v0.3.0 set" begin
        names = closed_function_names()
        @test "datetime.year" in names
        @test "datetime.month" in names
        @test "datetime.day" in names
        @test "datetime.hour" in names
        @test "datetime.minute" in names
        @test "datetime.second" in names
        @test "datetime.day_of_year" in names
        @test "datetime.julian_day" in names
        @test "datetime.is_leap_year" in names
        @test "interp.searchsorted" in names
        @test "interp.linear" in names
        @test "interp.bilinear" in names
        @test length(names) == 12
    end

    # Unknown name → diagnostic `unknown_closed_function`.
    @testset "Unknown name rejects with stable diagnostic code" begin
        err = try
            evaluate_closed_function("datetime.century", [0.0])
            nothing
        catch e
            e
        end
        @test err isa ClosedFunctionError
        @test (err::ClosedFunctionError).code == "unknown_closed_function"
    end
end
